### ISSUE TITLE
Loss of Precision in BatchNorm and output_var may be wrong

### ISSUE THREAD

**Issue by wkcn** [apache/incubator-mxnet#9216]
## Description
1. BatchNorm loses a little precision
2. the output_var in BatchNorm may be wrong

## Environment Info
OS: Arch Linux 4.14.8
MXNet: 1.0.0 and 1.0.1 (the latest version, CPU version)

## Build Config
make -j8 USE_OPENCV=1 USE_BLAS=openblas
***

Hi, there.

I converted [ResNet Model on Caffe](https://github.com/KaimingHe/deep-residual-networks) to [ResNet model on MXNet](https://github.com/wkcn/resnet-v1-mx).

And I found that the output results between Caffe and MXNet are different.

The reason is that the computations of Caffe and MXNet are different.

For the BatchNorm in Caffe, the output is `(x - mean(x)) / sqrt(var(x) + eps)`.

For the BatchNorm in MXNet, the output is `(x - mean(x)) * factor`, and `factor = 1.0 / sqrt(var(x) + eps)`.

I think the method in MXNet will **lose a little precision** but bring the **higher performance** (Reduce the times of division).

At the same time, I found that the `output_var` in BatchNorm may be wrong. 

The `output_var` is **invstd**, namely the multiplicative inverse of the standard deviation. I think it should be the variance.

## Steps to reproduce

Here is [my testing code](https://github.com/wkcn/test_mxnet_bn).

I compare three outputs:

- numpy (compute manally)
- caffe
- mxnet

```
caffe and numpy 0.0 0.0
caffe and mx 16.0 2.36527e-07
numpy and mx 16.0 2.36527e-07
```

The first column is the `maxmimum absolute error`, and the second column is the `maxmimum relative error`.

# What I have tried to solve it

I change the BatchNorm implement in MXNet, and the output is below:

```
caffe and numpy 0.0 0.0
caffe and mx 0.0 0.0
numpy and mx 0.0 0.0
```

The modified BatchNorm(cpu) code is [here](https://github.com/wkcn/incubator-mxnet/commit/5ecd4882bc043cf059e962f7ce488270bafa07c7).

**Comment by sxjscience** [comment]
What's the result if we use this line instead: https://github.com/wkcn/test_mxnet_bn/blob/master/compute.py#L15. I think it may be the same as the MXNet version. Also, the rel_error is relatively small and should not be a problem.

**Comment by wkcn** [comment]
@sxjscience If using the line 15th, the result will be the same as that of MXNet.

The rel_error is relatively small.

However, there may be **a bug**. 


The [documents](https://mxnet.incubator.apache.org/api/python/symbol/symbol.html?highlight=batchnorm#mxnet.symbol.BatchNorm) said that
output_mean_var (boolean, optional, default=0) – Output All,normal mean and var

https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cc#L147
The output_var is **invstd** now, I think it should be **the variance**.

**Comment by sxjscience** [comment]
Looks like a bug, I'm not sure. @piiswrong @cjolivier01

**Comment by chinakook** [comment]
Please be caution that CUDNN cannot handle BN when eps <= 1e-5. In this situation, you can set cudnn_off=true in BN of MXNet.

**Comment by wkcn** [comment]
@chinakook Thank you!

I wrote [a new code](https://github.com/wkcn/test_mxnet_bn/tree/ndarray) to test.
And I found that using CUDNN and not using CUDNN on Caffe get the same BN output.
But there is a little difference between the CPU result and the GPU result on Caffe.
```
max absolute error, max relative error, mean error
8.0, 3.4666334e-07, 0.024640804
```

The comparison between MXNet and Caffe is showed below.

Using the GPU result of Caffe.
```
('context: ', cpu(0))
('cudnn_off: ', True)
ndarray: y = gamma * [(x - mean) / sqrt(var + eps)] + beta

                max absolute error, max relative error, mean error
('caffe and ndarray', 8.0, 3.4666334e-07, 0.024640804)
('caffe and mx', 16.0, 3.4666334e-07, 0.057109512)
('ndarray and mx', 16.0, 2.3652711e-07, 0.038025968)
```

```
('context: ', gpu(0))
('cudnn_off: ', True)
ndarray: y = gamma * [(x - mean) / sqrt(var + eps)] + beta

                max absolute error, max relative error, mean error
('caffe and ndarray', 8.0, 3.4666334e-07, 0.024640804)
('caffe and mx', 16.0, 2.3644267e-07, 0.052587245)
('ndarray and mx', 16.0, 2.4033051e-07, 0.054628547)
```

```
('context: ', gpu(0))
('cudnn_off: ', False)
ndarray: y = gamma * [(x - mean) / sqrt(var + eps)] + beta

                max absolute error, max relative error, mean error
('caffe and ndarray', 8.0, 3.4666334e-07, 0.024640804)
('caffe and mx', 16.0, 2.3644267e-07, 0.052587245)
('ndarray and mx', 16.0, 2.4033051e-07, 0.054628547)
```

**Comment by sxjscience** [comment]
@wkcn Can you confirm that the output is invstd instead of var?

**Comment by sxjscience** [comment]
@wkcn These errors are too small and will cause no problem. However, there is certainly a bug if the output is invstd instead of var.

**Comment by wkcn** [comment]
@sxjscience 
Thank you!

I'm sure that the output_var is invstd instead of variance.
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/batch_norm.cc#L147

**Comment by piiswrong** [comment]
the documentation should be fixed

**Comment by sxjscience** [comment]
I find it will be improper to output the invstd as the value will be undefined when batch_size=1. We should consider to change it back to output the variance.

**Comment by cjolivier01** [comment]
Just FYI, CUDNN stores invstd for moving_variance rather than actual variance, if I remember correctly.
This can be seen in the unit test not checking for those being the same when CUDNN is enabled: https://github.com/apache/incubator-mxnet/blob/f9fd88bf178c0005a81ebf909750d5786bb1ec69/tests/cpp/operator/batchnorm_test.cc#L409

**Comment by sxjscience** [comment]
@cjolivier01 I see, we need to test the correctness of the mean and "var" as we can fetch these two values by turning on `output_mean_var`. https://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.BatchNorm

**Comment by sxjscience** [comment]
@cjolivier01 In that case, it should be okay to keep the invstd format. We need to test the case when `output_mean_var` is turned on.
